tests/net/mod.rs

```diff
@@ -990,24 +986,42 @@ where
 
 impl<C, D, N> VirtualNet<D>
 where
-    D: DistAlgorithm<Output = Batch<C, N>>,
+    D: DistAlgorithm<NodeId = N, Output = Batch<C, N>>,
     C: Contribution + Clone,
     N: NodeIdT,
 {
-    /// Verifies that all nodes' outputs agree, and returns the output.
-    pub fn verify_batches(&self) -> &[Batch<C, N>] {
-        let first = self.correct_nodes().nth(0).unwrap().outputs();
-        let pub_eq = |(b0, b1): (&Batch<C, _>, &Batch<C, _>)| b0.public_eq(b1);
-        for (i, node) in self.correct_nodes().enumerate().skip(0) {
+    /// Verifies that all nodes' outputs agree, given a correct "full" node that output all
+    /// batches in a total order and with no gaps.
+    ///
+    /// The output of the full node is used to derive in expected output of other nodes in every
+    /// epoch. After that the check ensures that correct nodes output the same batches in epochs
+    /// when those nodes were participants (either validators or candidates).
+    pub fn verify_batches<E>(&self, full_node: &Node<D>)
+    where
+        Batch<C, N>: SenderQueueableOutput<N, E>,
+    {
+        let mut participants: BTreeSet<N> = self.nodes().map(Node::id).cloned().collect();
+        let mut expected: BTreeMap<N, Vec<_>> = BTreeMap::new();
+        for batch in &full_node.outputs {
+            for id in &participants {
+                expected.entry(id.clone()).or_default().push(batch);
+            }
+            if let Some(new_participants) = batch.participant_change() {
+                participants = new_participants;
+            }
+        }
+        for node in self.correct_nodes().filter(|n| n.id() != full_node.id()) {
```

Comment:
Actually, maybe let's not even filter here, so we get alerted it we passed in a non-full node by accident?

Reply:
If there is another correct node with more batches, we will be alerted with the filter too! It happened to me a few times.